--- openrecoveryscript.cpp	2026-02-04 07:24:46.031963279 +0000
+++ openrecoveryscript.cpp	2026-02-04 07:24:46.031963279 +0000
@@ -1,4 +1,6 @@
 /*
+#include <string>
+#include <cerrno>
 	Copyright 2003 to 2017 TeamWin
 	This file is part of TWRP/TeamWin Recovery Project.
 
@@ -160,6 +162,14 @@
 				} else if (strcmp(value, "dalvik") == 0 || strcmp(value, "dalvick") == 0 || strcmp(value, "dalvikcache") == 0 || strcmp(value, "dalvickcache") == 0) {
 					PartitionManager.Wipe_Dalvik_Cache();
 				} else if (strcmp(value, "data") == 0 || strcmp(value, "/data") == 0 || strcmp(value, "factory") == 0 || strcmp(value, "factoryreset") == 0) {
+    gui_print("Forcing format data before wipe to avoid loop.\n"); 
+    PartitionManager.Format_Data(); 
+    if (PartitionManager.Wipe_By_Path("/data")) { 
+        gui_print("Wipe data success after format.\n"); 
+    } else { 
+        gui_print("Wipe data failed.\n"); 
+        ret_val = 1; 
+    }
 					PartitionManager.Factory_Reset();
 				} else {
 					LOGERR("Error with wipe command value: '%s'\n", value);
@@ -168,6 +178,14 @@
 			} else if (strcmp(command, "format") == 0) {
 				// Format
 				if (strcmp(value, "data") == 0 || strcmp(value, "/data") == 0 || strcmp(value, "factory") == 0 || strcmp(value, "factoryreset") == 0) {
+    gui_print("Forcing format data before wipe to avoid loop.\n"); 
+    PartitionManager.Format_Data(); 
+    if (PartitionManager.Wipe_By_Path("/data")) { 
+        gui_print("Wipe data success after format.\n"); 
+    } else { 
+        gui_print("Wipe data failed.\n"); 
+        ret_val = 1; 
+    }
 					PartitionManager.Format_Data();
 				} else {
 					LOGERR("Error with format command value: '%s'\n", value);
